lms/djangoapps/certificates/views/webview.py

```diff
@@ -588,46 +546,155 @@ def render_html_view(request, user_id, course_id):
+        context.update(course.cert_html_view_overrides)
 
-    # FINALLY, render appropriate certificate
-    return _render_certificate_template(request, context, course, user_certificate)
+        # Track certificate view events
+        _track_certificate_events(request, context, course, user, user_certificate)
+
+        # Render the certificate
+        return _render_valid_certificate(request, context, custom_template)
+
+
+def _get_catalog_data_for_course(course_key):
+    """
+    Retrieve data from the Discovery service necessary for rendering a certificate for a specific course.
+    """
+    course_certificate_settings = CertificateGenerationCourseSetting.get(course_key)
+    if not course_certificate_settings:
+        return {}
+
+    catalog_data = {}
+    course_run_fields = []
+    if course_certificate_settings.language_specific_templates_enabled:
+        course_run_fields.append('content_language')
+    if course_certificate_settings.include_hours_of_effort:
+        course_run_fields.extend(['weeks_to_complete', 'max_effort'])
+
+    if course_run_fields:
+        course_run_data = get_course_run_details(course_key, course_run_fields)
+        if course_run_data.get('weeks_to_complete') and course_run_data.get('max_effort'):
+            try:
+                weeks_to_complete = int(course_run_data['weeks_to_complete'])
+                max_effort = int(course_run_data['max_effort'])
+                catalog_data['hours_of_effort'] = weeks_to_complete * max_effort
+            except ValueError:
+                log.exception('Error occurred while parsing course run details')
+        catalog_data['content_language'] = course_run_data.get('content_language')
+
+    return catalog_data
+
+
+def _get_custom_template_and_language(course_id, course_mode, course_language):
+    """
+    Return the custom certificate template, if any, that should be rendered for the provided course/mode/language
+    combination, along with the language that should be used to render that template.
+    """
+    closest_released_language = _get_closest_released_language(course_language) if course_language else None
+    template = get_certificate_template(course_id, course_mode, closest_released_language)
+
+    if template and template.language:
+        return (template, closest_released_language)
+    elif template:
+        return (template, settings.LANGUAGE_CODE)
+    else:
+        return (None, None)
+
+
+def _get_closest_released_language(target):
+    """
+    Return the language code that most closely matches the target and is fully supported by the LMS, or None
+    if there are no fully supported languages that match the target.
+    """
+    match = None
+    languages = released_languages()
+
+    for language in languages:
+        if language.code == target:
+            match = language.code
+            break
+        elif (match is None) and (language.code[:2] == target[:2]):
```

Comment:
Just checking for clarity, this here will accept the first language that matches the 2 char code, unless later on the exact language code is found, then it will match that?

Reply:
Correct.